csharp/src/Ice/Reference.cs

```diff
@@ -584,6 +584,101 @@ static void StartQueryOption(StringBuilder sb, ref bool firstOption)
             }
         }
 
+        internal async ValueTask<Connection> GetConnectionAsync(CancellationToken cancel)
+        {
+            (Connection? connection, bool cached, IReadOnlyList<Endpoint> endpoints, List<Connector>? connectors) =
+                await GetExistingConnectionAsync(cancel).ConfigureAwait(false);
+
+            if (connection == null)
+            {
+                OutgoingConnectionFactory connectionFactory = Communicator.OutgoingConnectionFactory;
+                if (connectors == null)
+                {
+                    (cached, endpoints) = await ComputeEndpointsAsync(cancel).ConfigureAwait(false);
+                    connectors = await connectionFactory.ComputeConnectorsAsync(
+                        this,
+                        endpoints,
+                        ImmutableList<Connector>.Empty,
+                        cancel).ConfigureAwait(false);
+                }
+
+                Debug.Assert(connectors.Count > 0);
+                for (int i = 0; i < connectors.Count && connection == null; ++i)
+                {
+                    try
+                    {
+                        connection = await connectionFactory.CreateConnectionAsync(ConnectionId,
+                                                                                    connectors[i],
+                                                                                    cancel).ConfigureAwait(false);
+                        if (IsConnectionCached)
+                        {
+                            _connection = connection;
+                        }
+                    }
+                    catch (Exception)
+                    {
+                        if (i + 1 == connectors.Count)
+                        {
+                            throw;
+                        }
+                    }
+                }
+            }
+            Debug.Assert(connection != null);
+            return connection;
+        }
+
+        internal async ValueTask<(Connection? Connection, bool Cached, IReadOnlyList<Endpoint> Endpoints, List<Connector>? Connectors)> GetExistingConnectionAsync(CancellationToken cancel)
```

Comment:
I'm not convinced we should add this method... I understand the goal is to try to reduce a bit the code duplication between InvokeAsync/GetConnectionAsync but in the end I don't think it reduces the code duplication much.

Reply:
I removed this method